replace jgrapht with graph4j in PAG and CallGraph

sahilagichani14 opened the pull request.

sahilagichani14 committed:
replace jgrapht with graph4j

sahilagichani14 committed:
fix graph4j api's

swissiety commented on sootup.callgraph/src/main/java/sootup/callgraph/GraphBasedCallGraph.java:
```diff
@@ -75,15 +79,15 @@ protected GraphBasedCallGraph(
       return;
     }
-    graph.addVertex(vertex);
+    graph.addLabeledVertex(vertex);
```
correct me if I'm wrong - but can't we pass in the MethodSignature itself? so we don't even need the Vertex class, anymore.

sahilagichani14 replied:
Hey, to add custom classes, we can either add an int (id for each vertex) and a label (Vertex class). Another option was to have 2 HashMaps to map IDs with method signatures, as I couldn't find an API to get a Mapping between vertex ID and label.

sahilagichani14 replied:
@swissiety, one question though, for the graph4J lib, it supports only int/primitive types for edges and vertices. Therefore, if I map the method signature with a unique int, it wouldn't be a good approach and might even exceed the int max value. I just want to ask before switching to a different library.

Also, the tests that fail in the pipeline are green locally, idk whats wrong in the pipeline?

swissiety committed:
remove intermediary/legacy Vertex class

sahilagichani14 committed:
filter nonNull edges

swissiety commented:
I think would make sense to me to make use of a custom datastructure. We don't really benefit (quite the contrary: performance wise we sometimes iterate over all saved/joined neighbouring edges of a vertice-> O(neighbors) to find a single extends) from a general graph datastructure as we are basically just storing multiple trees within the same graph which has the same vertices in common.

swissiety commented:
@sahilagichani14 maybe it makes sense to extend existing graph4j structures to store those trees? especially making use of storing primitive integer values representing the vertices like graph4j does so that we can later on implement an efficient version of LCA (used in type inference) without the need for additional mappings etc - for lca see https://www.baeldung.com/cs/tree-lowest-common-ancestor

swissiety committed:
fix porting typehierarchy to graph4j as well

swissiety committed:
print debug

swissiety committed:
print debug

swissiety committed:
print debug

swissiety committed:
print debug

swissiety committed:
print debug

swissiety commented:
@Sahil lets merge this branch: i.e. please revert the failing changes from the typehierarchy - lets tackle that in another PR

swissiety committed:
rvert typehierarchy (for now)

swissiety committed:
style

swissiety committed:
fix doc: remove qilin - point to boomerang for pointer analysis

swissiety committed:
adapt PAG.java

swissiety committed:
style

swissiety committed:
remove legacy reference

swissiety committed:
revert using grah4j for Typehierarchy for now

swissiety committed:
fix dependencies

swissiety committed:
adapt tests

swissiety committed:
state

swissiety committed:
javadoc deployment is now handled in documentation

swissiety committed:
silence maven transfers

JonasKlauke committed:
Merge branch 'develop' into 1316-feature-replace-the-backing-graph-library-of-typehierarchy-and-callgraph-by-a-more-performant-choice

JonasKlauke committed:
Merge branch 'develop' into 1316-feature-replace-the-backing-graph-library-of-typehierarchy-and-callgraph-by-a-more-performant-choice

JonasKlauke committed:
simplified the multi call testcase

JonasKlauke committed:
simplified the multi call testcase

JonasKlauke committed:
fixed containsCall to take the actual label of the edge

JonasKlauke committed:
fixed containsCall to take the actual label of the edge

JonasKlauke committed:
removed bodyintercepters from callgraph to speedup

JonasKlauke committed:
removed bodyintercepters from callgraph to speedup

JonasKlauke committed:
fmt

JonasKlauke committed:
fmt

JonasKlauke commented:
The graph is not optimized and implemented to support Pseudograph.
The API does not support to return multiple edges betwenn the same vertexes.
Additonally, jgrapht has a bigger community. Graph4J is run by 2 developers.
I'm not sure if we should change. 
I think the overhead that is caused to use graph4j with multiple edges is bigger than the gain.
I would say we shoudl put this PR as not planned

JonasKlauke reviewed (changes requested):
multi edges support needed in graph4j

kadirayk committed:
Merge branch 'develop' into pta-wip

swissiety committed:
fix name suffix loop in createUniqueLocal to produce sequential suffixes

When multiple local variable slots share the same debug name from different
scopes, the disambiguation loop was appending to an already-mutated candidate
instead of the original base name. This produced 'x_1_2' instead of 'x_2'
for a third collision.

Commit:
use LocalVariableTable scope ranges when resolving local names

Pass the current instruction to determineLocalName so that slot reuse
across different scopes picks the right debug name instead of always
returning the first match in the LocalVariableTable.

Commit:
route NOP statement registration through setStmt

Use setStmt instead of directly writing to insnToStmt so NOP handling
goes through the same bookkeeping path as all other instruction handlers.

swissiety committed:
draft version immutable cfg

Commit:
replace O(n) replaceStmt scan with O(1) reverse map

Add stmtToInsn (Map<Stmt, AbstractInsnNode>) maintained in setStmt and
replaceStmt, eliminating the full entrySet iteration on every replacement.

Commit:
add tests for local name uniqueness and scope-aware naming

Test that three locals sharing the same debug name produce candidate,
candidate_1, candidate_2 — not the buggy candidate_1_2 — and that
slot-reused locals in different scopes get distinct names.

dependabot[bot] committed:
Bump github/codeql-action from 4.35.4 to 4.35.5

Bumps [github/codeql-action](https://github.com/github/codeql-action) from 4.35.4 to 4.35.5.
- [Release notes](https://github.com/github/codeql-action/releases)
- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/github/codeql-action/compare/68bde559dea0fdcac2102bfdf6230c5f70eb485e...9e0d7b8d25671d64c341c19c0152d693099fb5ba)

---
updated-dependencies:
- dependency-name: github/codeql-action
  dependency-version: 4.35.5
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

swissiety committed:
Merge pull request #1592 from soot-oss/dependabot/github_actions/github/codeql-action-4.35.5

Bump github/codeql-action from 4.35.4 to 4.35.5

swissiety committed:
flow analysis: O(1) array indexing replaces HashMaps

stmtToBeforeFlow/After → Object[] indexed via ImmutableBlockCFG.indexOf();
ordinalMap eliminated; Entry.ordinal set by Orderer, read by EntryBackedPriorityQueue.

swissiety committed:
fix MutableBlockControlFlowGraph; ensure RPO traversals / fallsthrough stmts are connected

dependabot[bot] committed:
Bump codecov/codecov-action from 6.0.0 to 6.0.1

Bumps [codecov/codecov-action](https://github.com/codecov/codecov-action) from 6.0.0 to 6.0.1.
- [Release notes](https://github.com/codecov/codecov-action/releases)
- [Changelog](https://github.com/codecov/codecov-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/codecov/codecov-action/compare/57e3a136b779b570ffcdbf80b3bdc90e7fab3de2...e79a6962e0d4c0c17b229090214935d2e33f8354)

---
updated-dependencies:
- dependency-name: codecov/codecov-action
  dependency-version: 6.0.1
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

swissiety committed:
Merge pull request #1594 from soot-oss/dependabot/github_actions/codecov/codecov-action-6.0.1

Bump codecov/codecov-action from 6.0.0 to 6.0.1

JonasKlauke committed:
Merge pull request #1587 from soot-oss/pta-wip

otf call graph and API using Sootup Values

dependabot[bot] committed:
Bump github/codeql-action from 4.35.5 to 4.36.0

Bumps [github/codeql-action](https://github.com/github/codeql-action) from 4.35.5 to 4.36.0.
- [Release notes](https://github.com/github/codeql-action/releases)
- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/github/codeql-action/compare/9e0d7b8d25671d64c341c19c0152d693099fb5ba...7211b7c8077ea37d8641b6271f6a365a22a5fbfa)

---
updated-dependencies:
- dependency-name: github/codeql-action
  dependency-version: 4.36.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

swissiety committed:
Merge pull request #1597 from soot-oss/dependabot/github_actions/github/codeql-action-4.36.0

Bump github/codeql-action from 4.35.5 to 4.36.0

dependabot[bot] committed:
Bump org.slf4j:slf4j-api from 2.0.17 to 2.0.18

Bumps org.slf4j:slf4j-api from 2.0.17 to 2.0.18.

---
updated-dependencies:
- dependency-name: org.slf4j:slf4j-api
  dependency-version: 2.0.18
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump org.slf4j:slf4j-simple from 2.0.17 to 2.0.18

Bumps org.slf4j:slf4j-simple from 2.0.17 to 2.0.18.

---
updated-dependencies:
- dependency-name: org.slf4j:slf4j-simple
  dependency-version: 2.0.18
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

swissiety committed:
Merge pull request #1598 from soot-oss/dependabot/maven/develop/org.slf4j-slf4j-api-2.0.18

Bump org.slf4j:slf4j-api from 2.0.17 to 2.0.18

swissiety committed:
Merge pull request #1599 from soot-oss/dependabot/maven/develop/org.slf4j-slf4j-simple-2.0.18

Bump org.slf4j:slf4j-simple from 2.0.17 to 2.0.18

swissiety committed:
Merge branch 'develop' into fix/asmmethodsource-improvements

Hazem-Gamall committed:
add line numbers

Hazem-Gamall committed:
format

dependabot[bot] committed:
Bump org.ow2.asm:asm-commons from 9.9.1 to 9.10

Bumps org.ow2.asm:asm-commons from 9.9.1 to 9.10.

---
updated-dependencies:
- dependency-name: org.ow2.asm:asm-commons
  dependency-version: '9.10'
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump org.ow2.asm:asm-util from 9.9.1 to 9.10

Bumps org.ow2.asm:asm-util from 9.9.1 to 9.10.

---
updated-dependencies:
- dependency-name: org.ow2.asm:asm-util
  dependency-version: '9.10'
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

swissiety committed:
Merge pull request #1602 from soot-oss/dependabot/maven/develop/org.ow2.asm-asm-util-9.10

Bump org.ow2.asm:asm-util from 9.9.1 to 9.10

swissiety committed:
Merge pull request #1601 from soot-oss/dependabot/maven/develop/org.ow2.asm-asm-commons-9.10

Bump org.ow2.asm:asm-commons from 9.9.1 to 9.10

swissiety committed:
Merge branch 'develop' into fix/asmmethodsource-improvements

swissiety committed:
Merge pull request #1600 from Hazem-Gamall/line-numbers-fix

add line numbers

dependabot[bot] committed:
Bump com.github.oshi:oshi-core from 6.12.0 to 7.0.1

Bumps [com.github.oshi:oshi-core](https://github.com/oshi/oshi) from 6.12.0 to 7.0.1.
- [Release notes](https://github.com/oshi/oshi/releases)
- [Changelog](https://github.com/oshi/oshi/blob/master/CHANGELOG.md)
- [Commits](https://github.com/oshi/oshi/compare/oshi-parent-6.12.0...oshi-parent-7.0.1)

---
updated-dependencies:
- dependency-name: com.github.oshi:oshi-core
  dependency-version: 7.0.1
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

swissiety committed:
Merge pull request #1595 from soot-oss/fix/asmmethodsource-improvements

fix LocalVariableName collisions/ranges and Instr/Stmt replacement speedup

swissiety committed:
dependency changed its api

swissiety committed:
Merge branch 'develop' into dependabot/maven/develop/com.github.oshi-oshi-core-7.0.1

swissiety committed:
Merge pull request #1603 from soot-oss/dependabot/maven/develop/com.github.oshi-oshi-core-7.0.1

Bump com.github.oshi:oshi-core from 6.12.0 to 7.0.1

swissiety committed:
fix GraphBasedCallGraph.containsCall()

swissiety committed:
Merge branch 'refs/heads/develop' into 1316-feature-replace-the-backing-graph-library-of-typehierarchy-and-callgraph-by-a-more-performant-choice

swissiety committed:
fix doc

swissiety committed:
Merge branch 'use_immutablecfg_in_flowanalysis' into 1316-feature-replace-the-backing-graph-library-of-typehierarchy-and-callgraph-by-a-more-performant-choice

swissiety committed:
remove more jgrapht from recent develop

swissiety committed:
fix UniverseSortedPriorityQueue

swissiety committed:
Merge remote-tracking branch 'origin/develop' into 1316-feature-replace-the-backing-graph-library-of-typehierarchy-and-callgraph-by-a-more-performant-choice

# Conflicts:
#	.github/workflows/codeql.yml
#	.github/workflows/scorecard.yml

swissiety committed:
adapt testcase

swissiety committed:
Merge branch 'develop' into 1316-feature-replace-the-backing-graph-library-of-typehierarchy-and-callgraph-by-a-more-performant-choice

# Conflicts:
#	sootup.java.bytecode.frontend/src/test/java/sootup/java/bytecode/frontend/interceptors/StaticSingleAssignmentFormerTest.java

swissiety committed:
adapt testcase

swissiety committed:
fix null handling - introduced by parallel edges (backing library did create right capacity, but did only copy one unique edge even when having multiple parallel edges)

swissiety committed:
break down cdg tests into smaller/more precise parts fixing non-determeinistic serialization of dotgraph node output
